ops/terraform/services/eft/lambda_src/sftp_outbound_transfer/sftp_outbound_transfer.py

```diff
@@ -0,0 +1,484 @@
+                )
+            )
+            sftp_priv_key = paramiko.RSAKey.from_private_key(
+                StringIO(global_config.sftp_user_priv_key)
+            )
+            ssh_client.get_host_keys().add(
+                hostname=global_config.sftp_hostname, keytype="ssh-rsa", key=sftp_host_key
+            )
+            ssh_client.connect(
+                global_config.sftp_hostname,
+                username=global_config.sftp_username,
+                pkey=sftp_priv_key,
+                look_for_keys=False,
+                allow_agent=False,
+                timeout=global_config.sftp_connect_timeout,
+            )
+        except (
+            BadHostKeyException,
+            AuthenticationException,
+            socket.error,
+            SSHException,
+            NoValidConnectionsError,
+        ):
+            logger.error(
+                "An unrecoverable exception occurred when attempting to connect to SFTP server %s",
+                global_config.sftp_hostname,
+                exc_info=True,
+            )
+            failed_full_path = "/".join(
+                [
+                    partner_config.failed_files_full_path,
+                    FileTransferError.SFTP_CONNECTION_ERROR,
+                    filename,
+                ]
+            )
+            logger.error(
+                'The file "%s" will be moved to %s in the %s bucket to indicate this failure',
+                decoded_file_key,
+                failed_full_path,
+                BUCKET,
+            )
+
+            try:
+                move_s3_object(object_key=decoded_file_key, destination=failed_full_path)
+            except RuntimeError:
+                logger.error(
+                    "Unrecoverable error occurred when attempting to move %s to %s",
+                    decoded_file_key,
+                    failed_full_path,
+                    exc_info=True,
+                )
+            return
+
+        logger.info(
+            "Connected successfully to %s. Attempting to upload %s to %s...",
+            global_config.sftp_hostname,
+            filename,
+            destination_folder,
+        )
+
+        try:
+            with ssh_client.open_sftp() as sftp_client:
+                current_dir = ""
+                for dir_part in destination_folder.split("/"):
+                    if not dir_part:
+                        continue
+                    current_dir += f"/{dir_part}"
+                    try:
+                        sftp_client.listdir(current_dir)
+                    except IOError:
+                        logger.info("%s does not exist, creating it", current_dir)
+                        sftp_client.mkdir(current_dir)
+
+                with sftp_client.open(f"{destination_folder}/{filename}", "wb", 32768) as f:
+                    s3_client.download_fileobj(Bucket=BUCKET, Key=decoded_file_key, Fileobj=f)  # type: ignore
```

Comment:
This ensures the Lambda does not have to download the file from S3 into a temp directory before uploading it. Instead, `paramiko` can open a file-like object on the SFTP Server, and we can tell `boto3` to download the object from S3 into that file-like object.

See [this StackOverflow Answer](https://stackoverflow.com/a/58725127) for more context.

Reply:
Using a fairly small Lambda (configured with 512 MB of memory) the Lambda was able to transfer a ~250 MB file in about 60 seconds, so this should be fairly quick. Note that was _within_ our AWS Account, so external transfers to the CMS EFT SFTP server _might_ be a bit slower.

Reply:
To clarify, it took 60 seconds to the Lambda to run _in total_. This includes the time it took to setup, transfer the file, and then execute the copy and delete (move) operations in S3. Being more specific:

- It took **40 seconds** for the Lambda to transfer the 260 MB file, starting from when the Lambda connected to the SFTP Server
- It took **8 seconds** for the Lambda to execute the copy and delete (move) S3 operations to move the 250 MB file from the pending directory to the sent directory